### PR TITLE
Deprecate the `vectorized` argument to pure_callback and ffi_call

### PR DESCRIPTION
**Pull request by dfm**
Currently, the primitives backing `pure_callback` and (by extension) `ffi_call` provide a default batching rule that is controlled by the parameter `vectorized` to these two functions. This batching rule (allegedly) implements the following behavior:

* For `vectorized=True`, vmapping results in a call to the callback directly with the batched inputs after moving the batch axis to the front. This assumes that the callback function directly handles vectorized inputs and (roughly) `callback(x_batch) == stack([callback(x) for x in x_batch])`.
* When `vectorized=False`, the vmap produces a sequential scan with the callback in the body, which could have significant performance implications.

While these don't seem like unreasonable defaults, the behavior when `vectorized=False` could result in unexpectedly poor performance for users who aren't familiar with this behavior. More importantly, the `vectorized=True` semantics are confusing—and I would argue wrong—when the function takes more than one input. For example, consider the following function:

```python
import jax
import jax.numpy as jnp

def callback(x, y):
  print(x.shape, y.shape)
  return x + y

def fun(x, y):
  shape = jnp.broadcast_shapes(x.shape, y.shape)
  dtype = jnp.result_type(x, y)
  out_type = jax.ShapeDtypeStruct(shape, dtype)
  return jax.pure_callback(callback, out_type, x, y, vectorized=True)
```

We can call this function under `vmap` as follows:

```python
jax.vmap(fun, in_axes=(0, None))(jnp.ones(3), 0.5)
```

This prints `(3,)` and `()` for the shapes, and the result has shape `(3,)`, as we expect. However, things get a little ugly if we apply a second `vmap` as follows:

```python
jax.vmap(jax.vmap(fun, in_axes=(0, None)), in_axes=(None, 0))(jnp.ones(3), jnp.ones(4))
```

In this case, by the time our callback is executed, it receives `x` and `y` with shapes `(3,)` and `(4,)`, respectively, and it crashes with a `TypeError` because of the mismatched shapes. Even worse, callback doesn't have enough metadata to gracefully handle this case no matter what; how can callback ever know which argument was batched first?

This behavior is bad, and the docs are (at least somewhat) misleading. This confusion has bitten me, and it has come up several times on the JAX issue tracker, most recently in #23624. In this PR, I propose a refactoring of this API.

## Proposal

I think that the best approach for handling these issues would be to deprecate both the default behavior (using a `lax.map`) and the `vectorized` parameter, in favor of an explicitly opt-in API using a `str` (or `Enum`) parameter called `vmap_method`. After the appropriate deprecation periods, my proposed API would have the following behavior:

* Calling `vmap` on a callback without an explicit `vmap_method` would raise an error.
* Using `vmap_method="sequential"` would produce a `lax.map` when vmapped (the current default behavior).
* Using `vmap_method="broadcast"` would be similar to the current `vectorized=True` behavior, but new axes would be added in the appropriate locations for unbatched inputs. For example, in the double `vmap` example from above, x would have shape `(1, 3)` and y would have shape `(4, 1)` when they reach the callback.
* Finally, `vmap_method="broadcast_fullrank"` would behave like broadcast, but the inputs would be tiled to the expected batched shape (e.g. `(4, 3)` in the example above). I would argue that the documentation for `pure_callback` and `ffi_call` currently imply this behavior for `vectorized=True`.

This proposed API would continue to support existing use cases, while minimizing the chance of confusion and performance footguns.

## Alternatives

Another approach would be to remove support for batching callbacks entirely in favor of an API like `custom_vmap` or something more general. This was what I suggested in response to #23624, but since `custom_vmap` isn't documented or well supported, I'm not sure this is a totally satisfying solution. Furthermore, I expect that the vast majority of use cases for batching callbacks will be supported by some simple rules like the ones listed above, with minimal maintenance burden. More advanced needs could still be covered by `custom_vmap`.

### CODE REVIEW

File: jax/_src/callback.py [hunk 1]
```diff
@@ -129,25 +135,50 @@ def callback_batching_rule(
...
 ):
-  axis_size = next(a.shape[d] for a, d in zip(args, dims)
-                   if d is not batching.not_mapped)
+  if isinstance(vectorized, DeprecatedArg) and vmap_method is None:
+    warnings.warn(
+        f"The default behavior of {prim.name} under vmap will soon "
+        "change. Currently, the default behavior is to generate a sequential "
+        "vmap (i.e. a loop), but in the future the default will be to raise "
+        "an error. To keep the current default, set vmap_method='sequential'.",
+        DeprecationWarning, stacklevel=2)
```
Comment (jakevdp) [inline jax/_src/callback.py:149]: We probably need a higher stacklevel to make sure this displays at the call-site.

Comment (dfm) [reply jax/_src/callback.py]: Very good point! It's definitely tricky to set the right value here in general. For example, the following works (i.e. the warning points to the `jax.pure_callback` line) with `stacklevel=6`:

```python
import jax
import jax.numpy as jnp
import numpy as np

def f(x, **kwargs):
  return jax.pure_callback(np.sin, x, x, **kwargs)

jax.vmap(f)(jnp.arange(4.0))
```

But, I can't find any stack level that works for:

```python
jax.vmap(jax.jit(f))(jnp.arange(4.0))
```

In this second case, `stacklevel=6` points to an `eval_jaxpr`, but more importantly, `jax.pure_callback` doesn't appear anywhere in the stack.

The best I was able to come up with to cover these two cases was a scheme that combines the traceback filtering logic from `traceback_util` and `catch_warnings` to re-emit the warnings (see my latest commit). With these changes, we emitted warnings for these two examples are:

```
demo.py:6: DeprecationWarning: The default behavior of pure_callback under vmap will soon change. Currently, the default behavior is to generate a sequential vmap (i.e. a loop), but in the future the default will be to raise an error. To keep the current default, set vmap_method='sequential'.
  return jax.pure_callback(np.sin, x, x, **kwargs)
```

and

```
demo.py:9: DeprecationWarning: The default behavior of pure_callback under vmap will soon change. Currently, the default behavior is to generate a sequential vmap (i.e. a loop), but in the future the default will be to raise an error. To keep the current default, set vmap_method='sequential'.
  jax.vmap(jax.jit(f))(jnp.arange(4.0))
```

respectively.

WDYT? Is there a better way? Should we just choose a level (e.g. `6`) and be done with it?

Comment (jakevdp) [reply jax/_src/callback.py]: Yeah that's tricky... I'd probably lean against walking the stack in order to dynamically set the stacklevel. Maybe just set to 6, even though that may be an underestimate of the true stack level in the case of nested transformations.

Comment (dfm) [reply jax/_src/callback.py]: Yeah, I think that's probably better too. Reverted and set to 6.

File: jax/_src/callback.py [hunk 2]
```diff
@@ -129,25 +135,50 @@ def callback_batching_rule(
...
-    vectorized: bool,
+    vectorized: bool | None | DeprecatedArg,
+    vmap_method: str | None,
     result_avals: Sequence[core.ShapedArray],
     **kwargs: Any,
 ):
-  axis_size = next(a.shape[d] for a, d in zip(args, dims)
-                   if d is not batching.not_mapped)
+  if isinstance(vectorized, DeprecatedArg) and vmap_method is None:
+    warnings.warn(
```
Comment (superbobry) [inline jax/_src/callback.py:144]: Should we add a TODO to drop this in 6 months?

Also, it would be nice to use `deprecations.warn` here so that we could accelerate the deprecation for Google-internal code.

File: jax/_src/callback.py [hunk 3]
```diff
@@ -129,25 +135,50 @@ def callback_batching_rule(
...
   new_args = [arg if dim is batching.not_mapped else
               batching.moveaxis(arg, dim, 0) for arg, dim in zip(args, dims)]
-  if vectorized:
-    result_avals = tuple(
-        core.unmapped_aval(axis_size, core.no_axis_name, 0, aval)  # type: ignore
-        for aval in result_avals)
+  batched_result_avals = tuple(
+      core.unmapped_aval(axis_size, core.no_axis_name, 0, aval)
+      for aval in result_avals)
+  if vmap_method == "legacy_vectorized":
```
Comment (superbobry) [inline jax/_src/callback.py:159]: Do we plan to remove this eventually?

**Review by superbobry — APPROVED** [review]

File: jax/_src/callback.py (hunk 2 above)
Comment (dfm) [reply jax/_src/callback.py]: Done and done. Thanks for the suggestions!

File: jax/_src/callback.py (hunk 3 above)
Comment (dfm) [reply jax/_src/callback.py:163]: I don't feel strongly about this. I don't think there's much maintenance burden to keeping this around for users who want to opt in to this behavior. I don't know why anyone would _want_ this, but I've been wrong in the past! :D

**Comment by terhorst** [comment]
Just a heads up, it seems this did not make it into 0.4.34 despite being mentioned in the changelog.

**Comment by dfm** [comment]
Thanks for the heads up! You're absolutely right, although the GitHub tag also doesn't seem to include it either. I'll update the changelog accordingly!

**Comment by froystig** [comment]
What does the "rank" in the name `"broadcast_fullrank"` refer to? It seems that the array ranks, as in `np.ndim`, are the same in either broadcast option.

What would you think of renaming the options here to:

* `"expand"` for current "broadcast" with singleton dimensions, and
* `"broadcast"` for current "broadcast_fullrank" with fully broadcast dimensions?

I'm trying to suggest `np.expand_dims` in the former case, since that inserts singleton dimensions, and to suggest `np.broadcast` in the latter case, since that behaves as a material common broadcast on all of its arguments.

**Comment by dfm** [comment]
Good points - naming is hard!

I like your suggestions, and definitely don't feel too attached to the names from this PR. Maybe `"broadcast"` by itself is confusing, since people have different intuitions for what broadcasting means. I think that the analogy to `expand_dims` is a good one, but `np.broadcast` doesn't seem like quite the right comparison here. Maybe `np.broadcast_arrays` is closer to the semantics we mean. What about `"expand_dims"` and `"broadcast_arrays"`?

**Comment by froystig** [comment]
Expanding to `"expand_dims"` sounds good to me.

And `"broadcast_arrays"`, or something like `"broadcast_all"`, sounds good to me.

**Comment by dfm** [comment]
@jakevdp, @superbobry — Do either of you have opinions about this naming question?
